### PR TITLE
Implement java.util.ArrayDeque.

### PR DESCRIPTION
**Pull request by andreaTP**

### CODE REVIEW

**Comment by sjrd** [comment]
Jenkins, test this please

**Comment by sjrd** [comment]
Could you directly write the tests as JUnit tests in `test-suite/shared/`, please? So that we also test that the JVM does what we think it does.

**Comment by andreaTP** [comment]
@sjrd I'm looking at the task to put the tests under shared, I really like it but, it needs to be done all around.

I think we can try to agree on this:
You can assign me on porting at least most of the javalib tests under shared in JUnit, but , to do this I will write a thin interface of compatibility
i.e. something like 
```scala
implicit class ToJUnitAssertFalse(b: => Boolean) {
  def toBeFalsy() = assertFalse(b)
}
```
but, for this PR we can try to go straight.

What do you think?

**Comment by andreaTP** [comment]
I think Jenkins failure is spurious here:
He says ```process out of memory```

**Comment by nicolasstucki** [comment]
Porting `ArrayDequeTest` to JUnit will not be possible without porting `AbstractCollectionTest` and all of its parents. This will be much more involved and I would leave that for the future in a separate PR.

What you should do for this PR is:
- Keep the current `ArrayDequeTest` in jasmine, but remove all the `it(...){...}` statements and rename it `ArrayDequeJSTest`. This will test the `AbstractCollection` API.
- Add a new `ArrayDequeTest` in JUnit in the `shared` directory where you add all the tests that you removed from the other one.

I've started working on porting all the existing `java.util.Collection` tests.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+  def push(e: E): Unit = addFirst(e)
+
+  def pop(): E = removeFirst()
+
+  def size(): Int = inner.size
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  private var status = 0L
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: Define the fields at the beginning of the files.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    offerFirst(e)
+
+  def addLast(e: E): Unit =
+    offerLast(e)
+
+  def offerFirst(e: E): Boolean = {
+    if (e == null) {
+      throw new NullPointerException()
+    } else {
+      inner.insert(0, e)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: This operation is linear and we require a constant time `offerFirst`.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      status += 1L
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: What if `status == Long.MaxValue`?

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+  def peekLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.last
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean = {
+    val index = inner.indexWhere(_ === o)
+    if (index >= 0) {
+      val res = inner.remove(index) != null
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: Here `res` will always be `true`.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+      if (res) status += 1L
+      res
+    } else
+      false
+  }
+
+  def removeLastOccurrence(o: Any): Boolean = {
+    val index = inner.lastIndexWhere(_ === o)
+    if (index >= 0) {
+      val res = inner.remove(index) != null
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: Same.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+  def peek(): E = peekFirst()
+
+  def push(e: E): Unit = addFirst(e)
+
+  def pop(): E = removeFirst()
+
+  def size(): Int = inner.size
+
+  override def isEmpty(): Boolean = inner.isEmpty
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: No need for the override, the version in `AbstractCollection` is good enough.

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 1 above)
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: Consider making it an `Int` for performance.

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        }
+      }
+    }
+  }
+
+  def iterator(): Iterator[E] =
+    _failFastIterator(inner.iterator)
+
+  def descendingIterator(): Iterator[E] =
+    _failFastIterator(inner.reverse.iterator)
```
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: This is not efficient. I would suggest implementing `_failFastIterator` based on direct access of `inner` using the index and abstract on the way the `index` is updated.

File: test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+object ArrayDequeTest extends ArrayDequeTest(new ArrayDequeFactory)
+
+abstract class ArrayDequeTest[F <: ArrayDequeFactory](listFactory: F)
+    extends AbstractCollectionTest(listFactory) {
+
+  override def testApi(): Unit = {
+
+    super.testApi()
+
+    it("add and remove properly in head and last positions") {
```
Comment (nicolasstucki) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: Most of these methods should be tested on a in `DequeTest`.

**Comment by nicolasstucki** [comment]
That is all

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 3 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: simply overflow

File: javalib/src/main/scala/java/util/ArrayDeque.scala [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+  def pop(): E = removeFirst()
+
+  def size(): Int = inner.size
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  private var status = 0L
+
+  private def _failFastIterator(iter: scala.collection.Iterator[E]) = {
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: No leading `_`

File: test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+      val diter = ll.descendingIterator()
+      for (i <- (0 until l.size()).reverse) {
+        expect(diter.hasNext()).toBeTruthy
+        expect(diter.next()).toEqual(s(i))
+      }
+      expect(diter.hasNext()).toBeFalsy
+    }
+  }
+}
+
```
Comment (nicolasstucki) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: You also need to create the `DequeFactory` with `def allFactories` method and add it to the `CollectionFactory`:
```scala
object DequeFactory {
  def allFactories: Iterator[CollectionFactory] =
    Iterator(new ArrayDequeFactory)
}
```
```scala
object CollectionFactory {
  def allFactories: Iterator[CollectionFactory] =
    ListFactory.allFactories ++ SetFactory.allFactories ++ DequeFactory.allFactories
}
```
This way the `ArrayDeque` will be tested against all tests of `CollectionsTest` as well.

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 1 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 7 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 6 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 2 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 4 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 5 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

File: javalib/src/main/scala/java/util/ArrayDeque.scala (hunk 9 above)
Comment (andreaTP) [inline javalib/src/main/scala/java/util/ArrayDeque.scala]: ok

**Comment by andreaTP** [comment]
I'm currently porting the tests to JUnit, not sure how to run the tests in JS and in jvm mode...

**Comment by sjrd** [comment]
> not sure how to run the tests in JS and in jvm mode...

Put it in `shared`. Then `testSuite/test` runs in js. `testSuiteJVM/test` runs in the JVM

File: test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala (hunk 8 above)
Comment (andreaTP) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: which ones?

**Comment by andreaTP** [comment]
tnx!

**Comment by andreaTP** [comment]
it looks like tests on floating point numbers behave differently on JS/JVM @nicolasstucki can you check the shared test in my last commit pleaz?

**Comment by andreaTP** [comment]
ok, probably just a missing check on the JS part, should I file the issue on the JUnit.js repo?

**Comment by andreaTP** [comment]
just tell me what I need to test better and I hope we are done.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+import org.junit.Test
+import org.junit.Assert._
+
+import scala.collection.JavaConversions._
+
+import java.util.ArrayDeque
+
+class ArrayDequeTest {
+
+  @Test def `should_add_and_remove_head_and_last`(): Unit = {
```
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: Given that you used `_` for the spaces and have no special characters you don't need the `s.

**Comment by nicolasstucki** [comment]
The tests look fine like that.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala (hunk 11 above)
Comment (andreaTP) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: I follow this style:
https://github.com/scala-js/scala-js/blob/master/test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/BooleanTest.scala#L15

please agree on one and let me know ;-)

Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: I made some major refactoring using regex expressions, I must have missed those. I'll clean up mine too, thanks ;)

Comment (andreaTP) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/ArrayDequeTest.scala]: ok after cleaning i'll go rebasing

**Comment by andreaTP** [comment]
here you are

**Comment by andreaTP** [comment]
am I supposed doing anything else on this?

**Comment by nicolasstucki** [comment]
LGTM.
Jenkins, retest this please.

**Comment by nicolasstucki** [comment]
Jenkins, test this please.

**Comment by nicolasstucki** [comment]
@andreaTP wait for #2076 to be merged and then rebase. This should fix the currenty failing test.

**Comment by andreaTP** [comment]
ok, waiting for you guys!

**Comment by sjrd** [comment]
You can now rebase on top of master.

**Comment by andreaTP** [comment]
done

**Comment by nicolasstucki** [comment]
Dear Jenkins, retest this please.

**Comment by nicolasstucki** [comment]
LGTM

**Pull request merged**